### PR TITLE
feat(utils): update eslint types

### PR DESCRIPTION
**Pull request by bradzacher**
This just updates eslint types to be in sync with the last few versions.
I just went through the [history on the nodejs api docs file](https://github.com/eslint/eslint/commits/f2c2d350425268efa4b78ee6e0a2df8860e0efad/docs/developer-guide/nodejs-api.md) and compared the commits to what we have.

### CODE REVIEW

File: packages/utils/src/ts-eslint/ESLint.ts [hunk 1]
```diff
@@ -338,6 +378,11 @@ namespace ESLint {
      * The method to convert the LintResult objects to text
      */
     format(results: LintResult[]): string;
+    /**
+     * The method to asynchronously convert the LintResult objects to text.
+     * @since 8.4.0
+     */
+    format(results: LintResult[]): Promise<string>;
```
Comment (Josh-Cena) [inline packages/utils/src/ts-eslint/ESLint.ts:385]: I could be wrong, but does this actually work? Can you have overloads that differ only by return types? My [test](https://www.typescriptlang.org/play?target=9#code/JYOwLgpgTgZghgYwgAgGIHsoFs5klZAbwChkzkB6AKitPOSuQBUALFLCMF9AE2THTIE6EADdoYfm2QAZUGABKEAM4BXADaT0AIwBWEBGGX9BkAB5g65KhSvIYmHGAAUUFRqMAuZMrBRQAOYA2gC6AJTevv4gAQDcdtS09AzM0hxcvCbIcMoAniAILFAi6KrK6rlCIuJQklwocuBKaprIOvqGxgL8EBYAdHYpAALKoEjIABx9ACx9AAyDNnYO2Liu7prKkX6BoRHIAArFWMDKEAA8UYEAfPFkAL7ExDwG6nBuVSC+9t4Yq3jQeLCL6SODIAC89j6KycziCACIAIzwgA0yHhACZ4eE+vUQM5nHAwhDrtkwkCRN9tBDsgB3ODASQwaGONYI5FozHY8lPXoAB0wkkI93iQA) suggests TS never picks up this overload

Comment (JoshuaKGoldberg) [reply packages/utils/src/ts-eslint/ESLint.ts:385]: Yeah this should apparently be a union: `string | Promise<string>`. Which I find weird but 🤷 compatibility.

https://github.com/eslint/eslint/issues/15242
https://github.com/eslint/eslint/pull/15243

Comment (bradzacher) [reply packages/utils/src/ts-eslint/ESLint.ts:385]: I was kind of hoping that TS would be smart here and merge the return value type.

I wanted to document the promise-based return separately as it is only accepted in a later version of ESLint.

But if TS won't let me, I'll just merge them together and describe it in the JSDoc 😢

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
LGTM - reviewed the ESLint blog posts for the versions and confirmed they all match up. Just need to change the overload to a union and it's good to go!

**Pull request merged**